lib/Haskell/Law/Integer.agda

```diff
@@ -0,0 +1,9 @@
+module Haskell.Law.Integer where
+
+open import Haskell.Prim
+
+pos-injective : ∀ { a b : Nat } → pos a ≡ pos b → a ≡ b
+pos-injective refl = refl
+
+neg-injective : ∀ { a b : Nat } → negsuc a ≡ negsuc b → a ≡ b
```

Comment:
I'm not sure if I understand correctly what you are suggesting. Do you mean this `Injective` should be some layer of abstraction for all these `f-injective` functions? Or a law that would be a counterpart to `cong : {A B : Set} → ∀ (f : A → B) {x y} → x ≡ y → f x ≡ f y` ? I don't think the latter would work, because I don't think it would be possible to prove such law for an unspecified f, and in case of the former, I'm not sure how to do it; could you give an example of what you have in mind?

Reply:
Alright, thanks for the explanation. I can definitely implement it for some functions, but I don't think it will work for all of them, for example the list injections ` {x y : a} {xs ys : List a} x ∷ xs ≡ y ∷ ys → xs ≡ ys` or Either ` ∀ {a b : Set} {z w : a} → (x y : Either a b) → ⦃ x ≡ Left z ⦄ → ⦃ y ≡ Left w ⦄ -> x ≡ y  → z ≡ w`.  Is it still useful to have this abstraction only for some injection proofs?

Reply:
The either case could be simplified to `Left z ≡ Left w → z ≡ w`, so can still be phrased as `Injective Left` (possibly with some type annotation for the right summand). 

The list case indeed does not fit, but the abstraction is still clearly useful to reduce duplication.

Reply:
The reason the Either case looks like that is because I couldn't make it work with the simpler signature, `: ∀ {a  : Set} {z w : a} →  (Left z ≡ Left w) -> z ≡ w` (or variations thereof), as Agda can't resolve some metavariables wrt the "inline" use of Left constructor in the type singature.

The type signature you gave for `Injective` also doesn't really work if there are other types in scope, like in the case of Either, Agda has troubles with resolving A and B metavariables for this simplified.

I'm sorry if I'm missing some obvious solutions here, I do not have that much experience with this particular area of Agda debugging ^^'

Reply:
> (possibly with some type annotation for the right summand).

```agda
data Either (a b : Set) : Set where
  Left  : a → Either a b
  Right : b → Either a b

private variable a b : Set

Injective : (a → b) → Set _
Injective f = ∀ {x y} → f x ≡ f y → x ≡ y

-- all the following work
postulate
  _ :  ∀ {z w : a} → (x y : Either a b) → ⦃ x ≡ Left z ⦄ → ⦃ y ≡ Left w ⦄ -> x ≡ y  → z ≡ w
  _ :  Injective (Left {a}{b})
  _ :  Injective {b = Either a b} Left
  _ :  Injective {a} (Left {b = b})
```

Reply:
In general, you can aid Agda's type inference by giving some implicit variables explicitly (using `{...}`).

Reply:
Thank you for the explanations! I managed to make it work.